### PR TITLE
Feature/fargate support v2

### PR DESCRIPTION
**Pull request by johnnorton**
The previous merge for ECS Fargate support did not include the ability to ENABLE "Assign Public IP" when creating the ECS Service.

From what I can tell, this is an option that only works for services on Fargate.  This option assigns a public IP to the ENI.  This is required to access the internet from the container if you do not have a NAT gateway setup.

### CODE REVIEW

**Comment by radeksimko** [comment]
Hi @johnnorton 
AFAICT the networking block is updatable per API docs:
http://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_UpdateService.html#ECS-UpdateService-request-networkConfiguration

Can you tell us more about the motivation behind the `ForceNew` in all fields?

Also `Elem` is used to define elements nested under the field, which is only relevant to `TypeList` or `TypeSet`, potentially `TypeMap`, but certainly not `TypeString`.

Finally we'll need to add a test to exercise this new field and ensure it works. Let me know if you need any help with that.

Thanks.

**Comment by johnnorton** [comment]
@radeksimko 
Thanks for the input.  My reasoning behind forcenew on networking was based on the Console not allowing updates to a service.  I will test with the CLI and see.

This is my first foray into GoLang.  I will will take a look at your suggestions and create a test.

Standby....

John

File: aws/resource_aws_ecs_service.go [hunk 1]
```diff
@@ -481,9 +494,8 @@ func resourceAwsEcsServiceUpdate(d *schema.ResourceData, meta interface{}) error
 		}
 	}
 
-	if d.HasChange("network_configration") {
-		input.NetworkConfiguration = expandEcsNetworkConfigration(d.Get("network_configuration").([]interface{}))
-	}
+	//d.HasChange("network_configration") is not working, so explicity calling method.
```
Comment (johnnorton) [inline aws/resource_aws_ecs_service.go]: Not sure why the the func HadChange is not detecting a change on "network_configuration".  Maybe we need to implement equals on the structure?

File: aws/resource_aws_ecs_service_test.go [hunk 2]
```diff
@@ -1261,6 +1261,7 @@ resource "aws_ecs_service" "main" {
 	network_configuration {
 		security_groups = ["${aws_security_group.allow_all_a.id}", "${aws_security_group.allow_all_b.id}"]
 		subnets = ["${aws_subnet.main.*.id}"]
+		assign_public_ip = "ENABLED"
```
Comment (johnnorton) [inline aws/resource_aws_ecs_service_test.go]: I feel like the test needs something more but I am not sure what else to do besides this....

**Comment by johnnorton** [comment]
@radeksimko.  Yes updating the network config does not need to force a new resource but the existing tasks will not get the new settings as per the docs.  

Please review my comments in the code and let me know.

Thanks!

File: aws/resource_aws_ecs_service.go (hunk 1 above)
Comment (johnnorton) [inline aws/resource_aws_ecs_service.go]: HasChange was not detecting a change here.  Maybe we need to implement the "equals" method?

File: aws/resource_aws_ecs_service_test.go (hunk 2 above)
Comment (johnnorton) [inline aws/resource_aws_ecs_service_test.go]: I am not sure if this is all I need to do for the test.  Seems like I am missing something

**Comment by johnnorton** [comment]
@radeksimko Anything else needed here?

**Comment by headlessme** [comment]
Would also like to see this merged

**Comment by dkuntz2** [comment]
Is there any news on getting this merged? Not being able to set auto assign public IP is preventing me from configuring fargate-backed ECS services in terraform.

As a note on the PR itself, I think it would be nicer to be able to use `true`/`false` rather than `ENABLED`/`DISABLED` as the value, given that this is just a boolean value and that feels like it's exposing too much of the underlying API in a way that isn't particularly helpful.

**Comment by johnnorton** [comment]
@radeksimko 

Made change suggested by @dkuntz2 dkuntz2

**Comment by austinkelleher** [comment]
This is also preventing me from configuring Fargate with Terraform. @radeksimko Can we please get this merged soon?

File: aws/resource_aws_ecs_service.go [hunk 3]
```diff
@@ -133,6 +133,11 @@ func resourceAwsEcsService() *schema.Resource {
 							Elem:     &schema.Schema{Type: schema.TypeString},
 							Set:      schema.HashString,
 						},
+						"assign_public_ip": &schema.Schema{
```
Comment (bflad) [inline aws/resource_aws_ecs_service.go]: You can remove `&schema.Schema` from this line here.

File: aws/resource_aws_ecs_service.go [hunk 4]
```diff
@@ -410,9 +415,14 @@ func flattenEcsNetworkConfigration(nc *ecs.NetworkConfiguration) []interface{} {
 	if nc == nil {
 		return nil
 	}
+
 	result := make(map[string]interface{})
 	result["security_groups"] = schema.NewSet(schema.HashString, flattenStringList(nc.AwsvpcConfiguration.SecurityGroups))
 	result["subnets"] = schema.NewSet(schema.HashString, flattenStringList(nc.AwsvpcConfiguration.Subnets))
+	result["assign_public_ip"] = "true"
+	if *nc.AwsvpcConfiguration.AssignPublicIp == "DISABLED" {
```
Comment (bflad) [inline aws/resource_aws_ecs_service.go]: The AWS SDK provides constants we can use here instead of directly testing our own strings (this applies for all "DISABLED"/"ENABLED" in this PR), e.g. `ecs.AssignPublicIpDisabled`/`ecs.AssignPublicIpEnabled`

File: website/docs/r/ecs_service.html.markdown [hunk 5]
```diff
@@ -102,6 +102,7 @@ Guide](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/cluster-query
 
 * `subnets` - (Required) The subnets associated with the task or service.
 * `security_groups` - (Optional) The security groups associated with the task or service. If you do not specify a security group, the default security group for the VPC is used.
+* `assign_public_ip` - (Optional) Valid values are "ENABLED" or "DISABLED".  Will assign a public IP address to the ENI.
```
Comment (bflad) [inline website/docs/r/ecs_service.html.markdown]: Documentation needs to be updated since the switch from string values to boolean values and should also note the default of `false`.

File: aws/resource_aws_ecs_service_test.go [hunk 6]
```diff
@@ -1517,6 +1517,7 @@ resource "aws_ecs_service" "main" {
 	network_configuration {
 		security_groups = ["${aws_security_group.allow_all_a.id}", "${aws_security_group.allow_all_b.id}"]
 		subnets = ["${aws_subnet.main.*.id}"]
+		assign_public_ip = true
```
Comment (bflad) [inline aws/resource_aws_ecs_service_test.go]: Can you please add an appropriate attribute test where this is used (and one in the other Fargate test where it should be the default of false)? e.g. `resource.TestCheckResourceAttr("aws_ecs_service.main", "network_configuration.0.assign_public_ip", "true")`

**Comment by johnnorton** [comment]
@bflad @radeksimko Sorry I didn't get on this sooner.  Learning Go :) . Anyway please review and see if we can get this merged.

**Comment by bflad** [comment]
I'll take another look in a few hours. Thanks for the updates and contributions while learning Go! 😃

**Comment by austinkelleher** [comment]
Thanks for your work on this @johnnorton. Looking forward to seeing this merged.

File: aws/resource_aws_ecs_service.go [hunk 7]
```diff
@@ -410,9 +415,14 @@ func flattenEcsNetworkConfigration(nc *ecs.NetworkConfiguration) []interface{} {
 	if nc == nil {
 		return nil
 	}
+
 	result := make(map[string]interface{})
 	result["security_groups"] = schema.NewSet(schema.HashString, flattenStringList(nc.AwsvpcConfiguration.SecurityGroups))
 	result["subnets"] = schema.NewSet(schema.HashString, flattenStringList(nc.AwsvpcConfiguration.Subnets))
+	result["assign_public_ip"] = "true"
+	if *nc.AwsvpcConfiguration.AssignPublicIp == ecs.AssignPublicIpDisabled {
```
Comment (bflad) [inline aws/resource_aws_ecs_service.go]: If `nc.AwsvpcConfiguration.AssignPublicIp` is `nil` (its listed as optional in the [ECS API documentation](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_AwsVpcConfiguration.html)), this can cause Terraform to crash. I can fix this (and clean it up a little since it will automatically default to `"false"`) prior to merge via:
```go
if nc.AwsvpcConfiguration.AssignPublicIp != nil {
	result["assign_public_ip"] = fmt.Sprintf("%v", *nc.AwsvpcConfiguration.AssignPublicIp == ecs.AssignPublicIpEnabled)
}
```

File: aws/resource_aws_lb_listener.go [hunk 8]
```diff
@@ -34,7 +34,6 @@ func resourceAwsLbListener() *schema.Resource {
 			"load_balancer_arn": {
 				Type:     schema.TypeString,
 				Required: true,
-				ForceNew: true,
```
Comment (bflad) [inline aws/resource_aws_lb_listener.go]: Unrelated change - will fix prior to merge

**Comment by bflad** [comment]
@johnnorton thank you so much for your contributions here! I will be pulling in your commits, but I am running into some issues acceptance testing this appropriately, which isn't really from your PR. I will fix those on top of this as it requires some refactoring of the resource/testing itself.

I wound up down this rabbit hole because ECS will return this error if you try to enable `assign_public_ip` on EC2 launch types:
`InvalidParameterException: Assign public IP is not supported for this launch type.`

Unfortunately, as noted in #85, the error handling in this resource is already causing trouble as it is retrying on some conditions like the above when it shouldn't be. The launch type and assign_public_ip combination is something we can validate at plan time, but the implementation is a little more complex than just few lines. I think it'll be best if I fix both of these issues together as its already hindering my acceptance testing. More soon.

**Comment by johnnorton** [comment]
@bflad Excellent.  Please let me know if I can help.

**Comment by cemo** [comment]
@bflad milestone seems `v1.10.0`, it should be `v1.9.0`

**Comment by bflad** [comment]
@cemo good catch, thanks and updated. Also, shout out to @johnnorton for working though the original PR.

**Comment by bflad** [comment]
This has been released in [terraform-provider-aws version 1.9.0](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#190-february-09-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
